### PR TITLE
STM32 FastIO using register access

### PR DESCRIPTION
**Pull request by ktand**
### Description

The HAL_STM32 is currently using `digitalWrite/digitalRead` for i/o-access. This is very slow. This PR implements "direct" register access for up to 65x better performance. There is no change to the `SET_INPUT/SET_OUTPUT` macros, they still uses the `pinMode` function.

Should be compatible with all STM32x-CPUs but has only been tested on STM32F407.

I have one question regarding the `SET_OUTPUT` macro as defined in the HAL_STM32:

`#define SET_OUTPUT(IO)          OUT_WRITE(IO, LOW)`

Is it a requirement to also set the pin LOW? @thinkyhead ?

**Note**: The delays for ST7920 has been modified to work with the faster io. 

### Comparison of the new vs old code:
With "direct" register access:

stm_fastwrite

Without...

stm_digitalwrite

Code used for measurements (compiled with -O2):

```
    while(1)
    {
        WRITE(PIN3, 0); WRITE(PIN2, 0); WRITE(PIN1, 0); WRITE(PIN0, 0);
        WRITE(PIN3, 0); WRITE(PIN2, 0); WRITE(PIN1, 0); WRITE(PIN0, 1);
        WRITE(PIN3, 0); WRITE(PIN2, 0); WRITE(PIN1, 1); WRITE(PIN0, 0);
        WRITE(PIN3, 0); WRITE(PIN2, 0); WRITE(PIN1, 1); WRITE(PIN0, 1);
        WRITE(PIN3, 0); WRITE(PIN2, 1); WRITE(PIN1, 0); WRITE(PIN0, 0);
        WRITE(PIN3, 0); WRITE(PIN2, 1); WRITE(PIN1, 0); WRITE(PIN0, 1);
        WRITE(PIN3, 0); WRITE(PIN2, 1); WRITE(PIN1, 1); WRITE(PIN0, 0);
        WRITE(PIN3, 0); WRITE(PIN2, 1); WRITE(PIN1, 1); WRITE(PIN0, 1);
        WRITE(PIN3, 1); WRITE(PIN2, 0); WRITE(PIN1, 0); WRITE(PIN0, 0);
        WRITE(PIN3, 1); WRITE(PIN2, 0); WRITE(PIN1, 0); WRITE(PIN0, 1);
        WRITE(PIN3, 1); WRITE(PIN2, 0); WRITE(PIN1, 1); WRITE(PIN0, 0);
        WRITE(PIN3, 1); WRITE(PIN2, 0); WRITE(PIN1, 1); WRITE(PIN0, 1);
        WRITE(PIN3, 1); WRITE(PIN2, 1); WRITE(PIN1, 0); WRITE(PIN0, 0);
        WRITE(PIN3, 1); WRITE(PIN2, 1); WRITE(PIN1, 0); WRITE(PIN0, 1);
        WRITE(PIN3, 1); WRITE(PIN2, 1); WRITE(PIN1, 1); WRITE(PIN0, 0);
        WRITE(PIN3, 1); WRITE(PIN2, 1); WRITE(PIN1, 1); WRITE(PIN0, 1);                        
    }
```

CC: @xC0000005 @hasenbanck
### Benefits

Improves i/o performance. Less delay between multiple pin writes as can be seen in the screenshots above.

### CODE REVIEW

File: Marlin/src/HAL/HAL_STM32/fastio_STM32.h [hunk 1]
```diff
@@ -24,28 +24,55 @@
...
+#define WRITE(IO,V)             _WRITE(IO,V)
+#define READ(IO)                _READ(IO)
+#define TOGGLE(IO)              _TOGGLE(IO)
+
 #define OUT_WRITE(IO,V)         do{ _SET_OUTPUT(IO); WRITE(IO,V); }while(0)
 
 #define SET_INPUT(IO)           _SET_MODE(IO, INPUT)                              /*!< Input Floating Mode                   */
 #define SET_INPUT_PULLUP(IO)    _SET_MODE(IO, INPUT_PULLUP)                       /*!< Input with Pull-up activation         */
 #define SET_INPUT_PULLDOWN(IO)  _SET_MODE(IO, INPUT_PULLDOWN)                     /*!< Input with Pull-down activation       */
 #define SET_OUTPUT(IO)          OUT_WRITE(IO, LOW)
```
Comment (AnHardt) [inline Marlin/src/HAL/HAL_STM32/fastio_STM32.h:76]: > SET_OUTPUT(IO) ... Is it a requirement to also set the pin LOW?

I'd say yes. A defined state is required. If you don't want a transition from input to LOW you could use OUT_WRITE(IO,HIGH).

Thanks.

File: Marlin/src/lcd/dogm/ultralcd_st7920_u8glib_rrd_AVR.cpp [hunk 2]
```diff
@@ -73,9 +73,9 @@
   #define CPU_ST7920_DELAY_2 DELAY_NS(0)
   #define CPU_ST7920_DELAY_3 DELAY_NS(189)
 #elif defined(ARDUINO_ARCH_STM32)
-  #define CPU_ST7920_DELAY_1 DELAY_NS(0)
-  #define CPU_ST7920_DELAY_2 DELAY_NS(0)
-  #define CPU_ST7920_DELAY_3 DELAY_NS(0)
+  #define CPU_ST7920_DELAY_1 DELAY_NS(300)
+  #define CPU_ST7920_DELAY_2 DELAY_NS(40)
+  #define CPU_ST7920_DELAY_3 DELAY_NS(340)
```
Comment (thinkyhead) [inline Marlin/src/lcd/dogm/ultralcd_st7920_u8glib_rrd_AVR.cpp:78]: These are surprisingly large delays. Are these the smallest that will work?

**Comment by ktand** [comment]
Yes. I tried to make them lower (by 40) but then the display got garbled.

I think the datasheet specifies a minimum clock cycle of 600ns, if I
remember correctly, so they aren't too much off.

Den tis 30 okt. 2018 23:59Scott Lahteine <notifications@github.com> skrev:

File: Marlin/src/lcd/dogm/ultralcd_st7920_u8glib_rrd_AVR.cpp (hunk 2 above)
Comment (AnHardt) [reply Marlin/src/lcd/dogm/ultralcd_st7920_u8glib_rrd_AVR.cpp:78]: That could be right and isn't surprising at all.
At the 16MHz AVRs one CPU cycle is ~62.5ns. I distributed the instructions to be part of the delay until i minimized the delays. (per ST7920 datasheet half a cycle in DELAY_3).
Here and now shifting, anding and pin setting contributes very little to the delays.

**Review by ktand — COMMENTED** [review]
The parentheses are incorrect below:

`#define _WRITE(IO, V) (FastIOPortMap[STM_PORT(digitalPin[IO])]->BSRR = _BV32(STM_PIN(digitalPin[IO] + (V ? 0 : 16))))`

should be

`#define _WRITE(IO, V) (FastIOPortMap[STM_PORT(digitalPin[IO])]->BSRR = _BV32(STM_PIN(digitalPin[IO]) + (V ? 0 : 16)))`

**Pull request merged**